beacon-chain/sync/initial-sync/blocks_fetcher.go

```diff
@@ -0,0 +1,400 @@
+package initialsync
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"math/rand"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/kevinms/leakybucket-go"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/pkg/errors"
+	eth "github.com/prysmaticlabs/ethereumapis/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/beacon-chain/blockchain"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
+	"github.com/prysmaticlabs/prysm/beacon-chain/p2p"
+	prysmsync "github.com/prysmaticlabs/prysm/beacon-chain/sync"
+	p2ppb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	"github.com/sirupsen/logrus"
+)
+
+var errNoPeersAvailable = errors.New("no peers available, waiting for reconnect")
+
+// blocksFetcherConfig is a config to setup the block fetcher.
+type blocksFetcherConfig struct {
+	ctx         context.Context
+	headFetcher blockchain.HeadFetcher
+	p2p         p2p.P2P
+	rateLimiter *leakybucket.Collector
+}
+
+// blocksFetcher is a service to fetch chain data from peers.
+// On an incoming requests, requested block range is evenly divided
+// among available peers (for fair network load distribution).
+type blocksFetcher struct {
+	ctx                    context.Context
+	headFetcher            blockchain.HeadFetcher
+	p2p                    p2p.P2P
+	rateLimiter            *leakybucket.Collector
+	requests               chan *fetchRequestParams   // incoming fetch requests from downstream clients
+	receivedFetchResponses chan *fetchRequestResponse // responses from peers are forwarded to downstream clients
+	quit                   chan struct{}              // termination notifier
+}
+
+// fetchRequestParams holds parameters necessary to schedule a fetch request.
+type fetchRequestParams struct {
+	ctx   context.Context // if provided, it is used instead of global fetcher's context
+	start uint64          // starting slot
+	count uint64          // how many slots to receive (fetcher may return fewer slots)
+}
+
+// fetchRequestResponse is a combined type to hold results of both successful executions and errors.
+// Valid usage pattern will be to check whether result's `err` is nil, before using `blocks`.
+type fetchRequestResponse struct {
+	params *fetchRequestParams
+	blocks []*eth.SignedBeaconBlock
+	err    error
+	peers  []peer.ID
+}
+
+// newBlocksFetcher creates ready to use fetcher.
+func newBlocksFetcher(cfg *blocksFetcherConfig) *blocksFetcher {
+	rateLimiter := cfg.rateLimiter
+	if rateLimiter == nil {
+		rateLimiter = leakybucket.NewCollector(allowedBlocksPerSecond, allowedBlocksPerSecond, false)
```

Comment:
Maybe add a comment on the boolean argument to NewCollector to signify what it means to the reader something like 
```go
leakybucket.NewCollector(..., false /* disable x */)
```